src/cdk-experimental/ui-patterns/behaviors/event-manager/event-manager.ts

```diff
@@ -0,0 +1,192 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+/**
+ * An event that supports modifier keys.
+ *
+ * Matches the native KeyboardEvent, MouseEvent, and TouchEvent.
+ */
+export interface EventWithModifiers extends Event {
+  ctrlKey: boolean;
+  shiftKey: boolean;
+  altKey: boolean;
+  metaKey: boolean;
+}
+
+/**
+ * Options that are applicable to all event handlers.
+ *
+ * This library has not yet had a need for stopPropagationImmediate.
+ */
+export interface EventHandlerOptions {
+  stopPropagation: boolean;
+  preventDefault: boolean;
+}
+
+/**
+ * A config that specifies how to handle a particular event.
+ */
+export interface EventHandlerConfig<T extends Event> extends EventHandlerOptions {
+  handler: (event: T) => Promise<boolean | void>;
+}
+
+/**
+ * Bit flag representation of the possible modifier keys that can be present on an event.
+ */
+export enum ModifierKey {
+  None = 0,
+  Ctrl = 0b1,
+  Shift = 0b10,
+  Alt = 0b100,
+  Meta = 0b1000,
+}
+
+/**
+ * Abstract base class for all event managers.
+ *
+ * Event managers are designed to normalize how event handlers are authored and create a safety net
+ * for common event handling gotchas like remembering to call preventDefault or stopPropagation.
+ */
+export abstract class EventManager<T extends Event> {
+  private _submanagers: EventManager<T>[] = [];
+
+  protected configs: EventHandlerConfig<T>[] = [];
+  protected beforeFns: ((event: T) => void)[] = [];
+  protected afterFns: ((event: T) => void)[] = [];
+
+  protected defaultHandlerOptions: EventHandlerOptions = {
+    preventDefault: false,
+    stopPropagation: false,
+  };
+
+  constructor(defaultHandlerOptions?: Partial<EventHandlerOptions>) {
+    this.defaultHandlerOptions = {
+      ...this.defaultHandlerOptions,
+      ...defaultHandlerOptions,
+    };
+  }
+
+  /**
+   * Composes together multiple event managers into a single event manager that delegates to the
+   * individual managers.
+   */
+  static compose<T extends Event>(...managers: EventManager<T>[]) {
```

Comment:
Having it in a static method will prevent the code from being tree shaken. Should we move it into a function?

Reply:
I removed this function as the `compose` functionality ended up not being used